tests/Medidata.ZipkinTracer.Core.Test/TraceProviderTests.cs

```diff
@@ -0,0 +1,222 @@
+                { TraceProvider.ParentSpanIdHeaderName, new [] { parentSpanId } },
+            });
+            var environment = new Dictionary<string, object>();
+
+            request.Stub(x => x.Headers).Return(headers);
+            context.Stub(x => x.Request).Return(request);
+            context.Stub(x => x.Environment).Return(environment);
+
+            var expectedIsSampled = fixture.Create<bool>();
+            var sampleFilter = MockRepository.GenerateStub<ZipkinSampler>(fixture.Create<string>(), fixture.Create<string>());
+            sampleFilter.Expect(x => x.ShouldBeSampled(context, null)).Return(expectedIsSampled);
+
+            // Act
+            var sut = new TraceProvider(sampleFilter, context);
+
+            // Assert
+            Assert.AreEqual(traceId, sut.TraceId);
+            Assert.AreEqual(spanId, sut.SpanId);
+            Assert.AreEqual(parentSpanId, sut.ParentSpanId);
+            Assert.AreEqual(expectedIsSampled, sut.IsSampled);
+        }
+
+        [TestMethod]
+        public void ConstructorWithContextHavingInvalidIdValues()
+        {
+            // Arrange
+            var fixture = new Fixture();
+            var traceId = fixture.Create<string>();
+            var spanId = fixture.Create<string>();
+            var parentSpanId = fixture.Create<string>();
+            var isSampled = fixture.Create<string>();
+
+            var context = MockRepository.GenerateStub<IOwinContext>();
+            var request = MockRepository.GenerateStub<IOwinRequest>();
+            var headers = new HeaderDictionary(new Dictionary<string, string[]>
+            {
+                { TraceProvider.TraceIdHeaderName, new [] { traceId } },
+                { TraceProvider.SpanIdHeaderName, new [] { spanId } },
+                { TraceProvider.ParentSpanIdHeaderName, new [] { parentSpanId } },
+                { TraceProvider.SampledHeaderName, new [] { isSampled } }
+            });
+            var environment = new Dictionary<string, object>();
+
+            request.Stub(x => x.Headers).Return(headers);
+            context.Stub(x => x.Request).Return(request);
+            context.Stub(x => x.Environment).Return(environment);
+
+            var expectedIsSampled = fixture.Create<bool>();
+            var sampleFilter = MockRepository.GenerateStub<ZipkinSampler>(fixture.Create<string>(), fixture.Create<string>());
+            sampleFilter.Expect(x => x.ShouldBeSampled(context, isSampled)).Return(expectedIsSampled);
+
+            // Act
+            var sut = new TraceProvider(sampleFilter, context);
+
+            // Assert
+            Assert.AreNotEqual(traceId, sut.TraceId);
+            Assert.AreEqual(sut.TraceId, sut.SpanId);
+            Assert.AreEqual(string.Empty, sut.ParentSpanId);
+            Assert.AreEqual(expectedIsSampled, sut.IsSampled);
+        }
+
+        [TestMethod]
+        public void ConstructorWithHavingTraceProviderInContext()
```

Comment:
This test is failing with `NullReferenceException` on my local machine because the mocked IOwinContext has a null Request on it.